libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -432,117 +420,99 @@ static inline void set_active(struct gic_dist_map *gic_dist, int irq, int value,
-     * not want to process any new overflow irqs */
-    size_t tail = lr_overflow->tail;
-    for (size_t i = lr_overflow->head; i != tail; i = LR_OF_NEXT(i)) {
-        if (vgic_vcpu_inject_irq(vgic_dist, vcpu, &lr_overflow->irqs[i]) == 0) {
-            lr_overflow->head = LR_OF_NEXT(i);
-            lr_overflow->full = (lr_overflow->head == LR_OF_NEXT(lr_overflow->tail));
-        } else {
-            break;
+    for (int i = 0; i < NUM_LIST_REGS; i++) {
+        if (vgic->lr_shadow[vcpu->vcpu_id][i] == NULL) {
+            return i;
         }
     }
-}
 
-static inline void vgic_handle_overflow(vgic_t *vgic, vm_vcpu_t *vcpu)
-{
-    vgic_handle_overflow_cpu(vgic, &vgic->lr_overflow[vcpu->vcpu_id], vcpu);
+    return -1;
 }
 
-static int vgic_vcpu_inject_irq(struct vgic_dist_device *d, vm_vcpu_t *inject_vcpu, struct virq_handle *irq)
+static int vgic_vcpu_load_list_reg(vgic_t *vgic, vm_vcpu_t *vcpu, int idx, struct virq_handle *irq)
 {
-    vgic_t *vgic;
-    int err;
-    int i;
-
-    vgic = vgic_device_get_vgic(d);
-
-    seL4_CPtr vcpu;
-    vcpu = inject_vcpu->vcpu.cptr;
-    for (i = 0; i < MAX_LR_OVERFLOW; i++) {
-        if (vgic->irq[inject_vcpu->vcpu_id][i] == NULL) {
-            break;
-        }
-    }
-    err = seL4_ARM_VCPU_InjectIRQ(vcpu, irq->virq, 0, 0, i);
-    assert((i < 4) || err);
-    if (!err) {
-        /* Shadow */
-        vgic->irq[inject_vcpu->vcpu_id][i] = irq;
+    int err = seL4_ARM_VCPU_InjectIRQ(vcpu->vcpu.cptr, irq->virq, 0, 0, idx);
+    if (err) {
+        ZF_LOGF("Failure loading vGIC list register");
         return err;
-    } else {
-        /* Add to overflow list */
-        return vgic_add_overflow(vgic, irq, inject_vcpu);
     }
+
+    vgic->lr_shadow[vcpu->vcpu_id][idx] = irq;
+
+    return 0;
 }
 
 int handle_vgic_maintenance(vm_vcpu_t *vcpu, int idx)
 {
     /* STATE d) */
-    struct gic_dist_map *gic_dist;
-    struct virq_handle **lr;
-
     assert(vgic_dist);
-    gic_dist = vgic_priv_get_dist(vgic_dist);
-    lr = vgic_priv_get_lr(vgic_dist, vcpu);
-    assert(lr[idx]);
+    struct gic_dist_map *gic_dist = vgic_priv_get_dist(vgic_dist);
+    vgic_t *vgic = vgic_device_get_vgic(vgic_dist);
+    struct virq_handle **lr_shadow = vgic->lr_shadow[vcpu->vcpu_id];
+    assert(lr_shadow[idx]);
 
     /* Clear pending */
-    DIRQ("Maintenance IRQ %d\n", lr[idx]->virq);
-    set_pending(gic_dist, lr[idx]->virq, false, vcpu->vcpu_id);
-    virq_ack(vcpu, lr[idx]);
+    DIRQ("Maintenance IRQ %d\n", lr_shadow[idx]->virq);
+    set_pending(gic_dist, lr_shadow[idx]->virq, false, vcpu->vcpu_id);
+    virq_ack(vcpu, lr_shadow[idx]);
+    lr_shadow[idx] = NULL;
 
     /* Check the overflow list for pending IRQs */
-    lr[idx] = NULL;
-    vgic_t *vgic = vgic_device_get_vgic(vgic_dist);
-    vgic_handle_overflow(vgic, vcpu);
+    struct virq_handle *virq = vgic_irq_dequeue(vgic, vcpu);
+    if (virq) {
+        return vgic_vcpu_load_list_reg(vgic, vcpu, idx, virq);
```

Comment:
I wonder, if we should print an error message here, because this would indicate a serious problem in the maintenance operation

```suggestion
        int err = vgic_vcpu_load_list_reg(vgic, vcpu, idx, virq);
        if (err) {
            ZF_LOGE("IRQ injection failed, code %d\n", err);
        }
        return err
```

Reply:
But `vgic_vcpu_load_list_reg()` will print an error message regardless who calls it?

Reply:
yes, you are right. It would make a nice error stack though seeing this bubble up though the layers then.

Reply:
Now that is something I can agree.